README.md
# PostCSS Uniapp Tailwindcss

[PostCSS] 在uniapp中使用tailwindcss兼容postcss7

[postcss]: https://github.com/postcss/postcss

```css
/* Input example */

.foo {
  width: 2px;
  height: 1rem;
}

.w-1\/2,
.w-0\.5 {
}
```

```css
/* Output example */
.foo {
  width: 4rpx;
  height: 32rpx;
}

.w-1_2,
.w-0-5 {
}
```

## Usage

**Step 1:** Install plugin:

```sh
npm install --save-dev postcss postcss-uniapp-tailwindcss-compatible
```

**Step 2:** Check you project for existed PostCSS config: `postcss.config.js`
in the project root, `"postcss"` section in `package.json`
or `postcss` in bundle config.

If you do not use PostCSS, add it according to [official docs]
and set this plugin in settings.

**Step 3:** Add the plugin to plugins list:

```diff
module.exports = {
  plugins: [
+   require('tailwindcss'),
+   require('postcss-uniapp-tailwindcss-compatible'),
    require('autoprefixer')
  ]
}
```

### Options

**default options:**

```js
{
  platform: 'mp',
  units: {
    px2rem: 0.25,
    px2rpx: 2,
    rem2rpx: 32,
    px2pt: 0.22,
    rpx2pt: 0.75,
    rem2pt: 28.125,
    'vw2%': 1,
    'vh2%': 1
  },
  unitIgnore: {
    selector: [],
    prop: [/^border-.*?width$/],
    value: []
  },
  unitGroup: {
    h5: ['px2rem'],
    mp: ['px2rpx', 'rem2rpx'],
    native: ['px2pt', 'rem2pt', 'rpx2pt', 'vw2%', 'vh2%']
  },
  selectorRules: [
    {
      pattern:
        /(\.[\S]+?(?<!>))\s?>\s?:not\(\[(hidden|template)\]\)\s?~\s?:not\(\[(hidden|template)\]\)/g,
      replacement: '$1 > view + view',
    },
    // .2xl
    {
      pattern: /\.\\/g,
      replacement: '._',
    },
    // *
    {
      pattern: /^\*$/,
      replacement: 'page',
    },
    // hover:w-1
    {
      pattern: /\\:/g,
      replacement: '_',
    },
    // w-1/2
    {
      pattern: /\\\//g,
      replacement: '_',
    },
    // w-0.5
    {
      pattern: /\\\./g,
      replacement: '-',
    },
  ]
}
```

| option                      | type          | description                                  |
| --------------------------- | ------------- | -------------------------------------------- |
| units                       | object        | unit conversion scale                        |
| unitIgnore                  | object        | Do not convert unit rules                    |
| unitIgnore.selector[]       | regex\|string | Do not convert declaration's parent selector |
| unitIgnore.prop[]           | regex\|string | Do not convert declaration's prop            |
| unitIgnore.value[]          | regex\|string | Do not convert declaration's value           |
| unitGroup                   | object        | unit group by platform                       |
| selectorRules               | array         | selector replace rules                       |
| selectorRules[].pattern     | regex\|string | selector replace pattern                     |
| selectorRules[].replacement | string        | selector replace replacement                 |

[official docs]: https://github.com/postcss/postcss#usage

# thanks

- https://github.com/LinXueYi/postcss-uniapp-tailwindcss


CHANGELOG.md
# Change Log

## 1.0.0 - 2021-11-26

第一个版本


v7.js
const postcss = require('postcss');
const config = require('./config');

module.exports = postcss.plugin(
  'postcss-uniapp-tailwindcss-compatible',
  (opts = {}) => {
    opts = { ...config, ...opts };
    const covertRules = opts.unitGroup[opts.platform].map((item) => {
      const rule = item.split('2');
      if (rule[0] && rule[1]) {
        return {
          key: rule[0],
          covert: rule[1],
        };
      }
    });

    const isRegExp = (val) => typeof val === 'object' && val instanceof RegExp;
    const isUnitIgnoreSelector = opts.unitIgnore
      && Array.isArray(opts.unitIgnore.selector)
      && opts.unitIgnore.selector.length > 0;
    const isUnitIgnoreProp = opts.unitIgnore
      && Array.isArray(opts.unitIgnore.prop)
      && opts.unitIgnore.prop.length > 0;
    const isUnitIgnoreValue = opts.unitIgnore
      && Array.isArray(opts.unitIgnore.value)
      && opts.unitIgnore.value.length > 0;
    const isSelectorRules = opts.selectorRules
      && Array.isArray(opts.selectorRules)
      && opts.selectorRules.length > 0;
    // Work with options here
    return (root) => {
      if (!isSelectorRules) return;

      root.walkRules((rule) => {
        // Transform CSS AST here
        opts.selectorRules.forEach((item) => {
          if (
            (typeof item.pattern === 'string'
              && rule.selector === item.pattern)
            || (isRegExp(item.pattern) && item.pattern.test(rule.selector))
          ) {
            rule.selector = rule.selector.replace(
              item.pattern,
              item.replacement,
            );
          }
        });

        rule.walkDecls((decl) => {
          covertRules.forEach((item) => {
            const times = opts.units[`${item.key}2${item.covert}`];
            if (!times || typeof times !== 'number') {
              return;
            }
            if (isUnitIgnoreSelector) {
              const selector = decl.parent.selector;
              const isIgnoreSelector = opts.unitIgnore.selector.some((item) => (
                (typeof item === 'string' && item === selector)
                  || (isRegExp(item) && item.test(selector))
              ));
              if (isIgnoreSelector) {
                return;
              }
            }
            if (isUnitIgnoreProp) {
              const isIgnoreProp = opts.unitIgnore.prop.some((item) => (
                (typeof item === 'string' && item === decl.prop)
                  || (isRegExp(item) && item.test(decl.prop))
              ));
              if (isIgnoreProp) {
                return;
              }
            }
            const regExp = new RegExp(`([\\d\\.]+)${item.key}`, 'ig');
            if (regExp.test(decl.value)) {
              const match = decl.value.match(regExp);
              match.forEach((matchItem) => {
                let val = matchItem.toLowerCase();
                if (isUnitIgnoreValue) {
                  const isIgnoreValue = opts.unitIgnore.value.some((item) => (
                    (typeof item === 'string'
                        && item.toLowerCase() === val)
                      || (isRegExp(item) && item.test(val))
                  ));
                  if (isIgnoreValue) {
                    return;
                  }
                }
                val = val.replace(item.key, '');
                const newVal = `${val * times}${item.covert}`;
                decl.value = decl.value.replace(matchItem, newVal);
              });
            }
          });
        });
      });
    };
  },
);


v8.js
const config = require('./config');

module.exports = (opts = {}) => {
  opts = { ...config, ...opts };
  const covertRules = opts.unitGroup[opts.platform].map((item) => {
    const rule = item.split('2');
    if (rule[0] && rule[1]) {
      return {
        key: rule[0],
        covert: rule[1],
      };
    }
  });

  const isRegExp = (val) => typeof val === 'object' && val instanceof RegExp;
  const isUnitIgnoreSelector = opts.unitIgnore
    && Array.isArray(opts.unitIgnore.selector)
    && opts.unitIgnore.selector.length > 0;
  const isUnitIgnoreProp = opts.unitIgnore
    && Array.isArray(opts.unitIgnore.prop)
    && opts.unitIgnore.prop.length > 0;
  const isUnitIgnoreValue = opts.unitIgnore
    && Array.isArray(opts.unitIgnore.value)
    && opts.unitIgnore.value.length > 0;
  const isSelectorRules = opts.selectorRules
    && Array.isArray(opts.selectorRules)
    && opts.selectorRules.length > 0;

  // Work with options here
  return {
    postcssPlugin: 'postcss-uniapp-tailwindcss',
    Rule(rule, postcss) {
      // Transform CSS AST here
      if (isSelectorRules) {
        opts.selectorRules.forEach((item) => {
          if (
            (typeof item.pattern === 'string'
              && rule.selector === item.pattern)
            || (isRegExp(item.pattern) && item.pattern.test(rule.selector))
          ) {
            rule.selector = rule.selector.replace(
              item.pattern,
              item.replacement,
            );
          }
        });
      }
    },
    Once(root, postcss) {
      root.walkDecls((decl) => {
        covertRules.forEach((item) => {
          const times = opts.units[`${item.key}2${item.covert}`];
          if (!times || typeof times !== 'number') {
            return;
          }
          if (isUnitIgnoreSelector) {
            const selector = decl.parent.selector;
            const isIgnoreSelector = opts.unitIgnore.selector.some((item) => (
              (typeof item === 'string' && item === selector)
                || (isRegExp(item) && item.test(selector))
            ));
            if (isIgnoreSelector) {
              return;
            }
          }
          if (isUnitIgnoreProp) {
            const isIgnoreProp = opts.unitIgnore.prop.some((item) => (
              (typeof item === 'string' && item === decl.prop)
                || (isRegExp(item) && item.test(decl.prop))
            ));
            if (isIgnoreProp) {
              return;
            }
          }
          const regExp = new RegExp(`([\\d\\.]+)${item.key}`, 'ig');
          if (regExp.test(decl.value)) {
            const match = decl.value.match(regExp);
            match.forEach((matchItem) => {
              let val = matchItem.toLowerCase();
              if (isUnitIgnoreValue) {
                const isIgnoreValue = opts.unitIgnore.value.some((item) => (
                  (typeof item === 'string' && item.toLowerCase() === val)
                    || (isRegExp(item) && item.test(val))
                ));
                if (isIgnoreValue) {
                  return;
                }
              }
              val = val.replace(item.key, '');
              const newVal = `${val * times}${item.covert}`;
              decl.value = decl.value.replace(matchItem, newVal);
            });
          }
        });
      });
    },
  };
};
module.exports.postcss = true;


config/index.js
module.exports = {
  platform: "mp",
  units: {
    px2rem: 0.25,
    px2rpx: 2,
    rem2rpx: 32,
    px2pt: 0.22,
    rpx2pt: 0.75,
    rem2pt: 28.125,
    "vw2%": 1,
    "vh2%": 1,
  },
  unitIgnore: {
    selector: [],
    prop: [/^border-.*?width$/],
    value: [],
  },
  unitGroup: {
    h5: ["px2rem"],
    mp: ["px2rpx", "rem2rpx"],
    native: ["px2pt", "rem2pt", "rpx2pt", "vw2%", "vh2%"],
  },
  selectorRules: [
    {
      pattern:
        /(\.[\S]+?(?<!>))\s?>\s?:not\(\[(hidden|template)\]\)\s?~\s?:not\(\[(hidden|template)\]\)/g,
      replacement: "$1 > view + view",
    },
    // .2xl
    {
      pattern: /\.\\/g,
      replacement: "._",
    },
    // *
    {
      pattern: /^\*$/,
      replacement: "page",
    },
    // hover:w-1
    {
      pattern: /\\:/g,
      replacement: "_",
    },
    // w-1/2
    {
      pattern: /\\\//g,
      replacement: "_",
    },
    // w-0.5
    {
      pattern: /\\\./g,
      replacement: "-",
    },
  ],
};
